Minimal implementation of the report system

lost-in-action opened the pull request:
From issue #102

Included most of the features from the issue. Since it's stateless (uses existing Discord message for finding and updating reports), the code is a bit messy and there may be a few issues.
I didn't introduce the feature that hides and then reinstate the reported message, I'm not sure if that would be what same actions as flagging means.
Currently there are 2 buttons for moderators: *Approve* and *Reject*, they only notify users who reported.
You can choose up to 3 rules (can be increased), but there are no implementation yet for non-community servers.

Users who will start reporting will have to wait 5 seconds before being able to report, during this time they are supposed to read the warning text. Afterwards the user can choose one or more rules from the dropdown list.

(Requires a lot of testing)

Jasius commented on src/db/index.js:
```diff
@@ -283,5 +285,40 @@ module.exports = {
+      console.log(result);
+    } catch(e) {
+      console.error(e);
+      throw 'Failed to block user from the report system';
```
```suggestion
      throw 'Failed to block the user.';
```

Jasius commented on src/db/index.js:
```diff
@@ -283,5 +285,40 @@ module.exports = {
+      console.log(result);
+    } catch(e) {
+      console.error(e);
+      throw 'Failed to unblock user from the report system';
```
```suggestion
      throw 'Failed to unblock the user.';
```

Jasius commented on src/features/commands/buttons/modMessageApproveReport.js:
```diff
@@ -0,0 +1,47 @@
+async function sendApprovalDM(user) {
+    const DM = await user.createDM();
+    await DM.send({
+        content: 'The moderation team has reviewed your report and took appropriate actions. The reported content was found to be against the server rules, we thank you for your report'
```
```suggestion
        content: 'Thank you for the report! The moderation team has reviewed your report and took action.'
```

Jasius commented on src/features/commands/buttons/modMessageRejectReport.js:
```diff
@@ -0,0 +1,48 @@
+async function sendRejectDM(user) {
+    const DM = await user.createDM();
+    await DM.send({
+        content: 'Following your report, the moderation team has estimated that the message is not against the server rules.\n'
+            + 'If you think this is an error, please contact a moderator through the moderation mail.'
```
```suggestion
        content: 'Thank you for your report! The moderation team has looked into the report and didn\'t take any action.\n'
            + 'If you think this is an error, please contact the moderation team.'
```

Jasius commented on src/features/commands/report.js:
```diff
@@ -0,0 +1,340 @@
+
+const alreadyReportedText = 'You already reported this message.';
+const alreadyReviewedText = 'This report was already reviewed by a moderator.';
+const blockedReportingText = `You are prevented from reporting, this could be because you abused the system.\n
```
```suggestion
const blockedReportingText = `You are blocked from reporting, this could be because you abused the system.\n
```

Jasius commented on src/features/commands/report.js:
```diff
@@ -0,0 +1,340 @@
+const alreadyReviewedText = 'This report was already reviewed by a moderator.';
+const blockedReportingText = `You are prevented from reporting, this could be because you abused the system.\n
+If you want more info, or if you think this is an error, contact moderators.`;
+const warnText = 'You are about to report a message. Abusing the report system coud lead to punishment. If this is really your intent you can go ahead.';
```
```suggestion
const warnText = 'You are about to report a message. Inappropriate use of reporting system may lead to punishment.';
```

Jasius commented on src/features/commands/report.js:
```diff
@@ -0,0 +1,340 @@
+const blockedReportingText = `You are prevented from reporting, this could be because you abused the system.\n
+If you want more info, or if you think this is an error, contact moderators.`;
+const warnText = 'You are about to report a message. Abusing the report system coud lead to punishment. If this is really your intent you can go ahead.';
+const genericReportText = 'Moderators will now review your report. Please note that they are not robots: it might take some time before an action is taken. You will receive a DM once moderators review your report, make sure you have enabled them for this server!';
```
```suggestion
const genericReportText = 'Report is pending a review, please note that this might take some time. You will receive a direct message once moderators review your report, make sure you have enabled them for this server!';
```

Jasius commented on src/features/commands/reportMod.js:
```diff
@@ -0,0 +1,52 @@
+                        .setDescription('User to block')
+                        .setRequired(true))),
+
+    async execute(client, interaction) {
+        switch(interaction.options.getSubcommand()) {
+            case 'block': {
+                const user = interaction.options.getUser('user');
+    
+                await db.saveBlockedReportUser(user.id, user.username);
+        
+                interaction.reply({
+                    content: `:exclamation: The user ${user} is now blocked from using report commands.`,
+                    ephemeral: true
+                })
+            } break;
+            case 'unblock': {
+                const user = interaction.options.getUser('user');
+    
+                await db.deleteBlockedReportUser(user.id);
+        
+                interaction.reply({
+                    content: `:white_check_mark: The user ${user} can now use report commands.`,
+                    ephemeral: true
+                })
+            } break;
+        }
+    }
+}
```
We should have permission checks here, to make sure that everyone isn't able to use these commands.

lost-in-action replied:
Yep I didn't bother trying to figure out how to check for permissions, but like in index.js, I think `message.member.permissions.serialize().KICK_MEMBERS` should be enough to check whether or not an user is a mod, or maybe eventually check if the user has access to the report channel

Jasius replied:
> I think `message.member.permissions.serialize().KICK_MEMBERS` should be enough to check whether or not an user is a mod

Certainly more than enough, as for checking whether mod has access to the channel seems kind of pointless as they wouldn't be able to interact anyways if they can't access the channel.

Jasius commented on src/index.js:
```diff
@@ -216,7 +216,7 @@ client.on('messageCreate', message => {
     })
   })
-  if (server.isPremium && !(message.member.permissions.serialize().KICK_MEMBERS || message.member.permissions.serialize().BAN_MEMBERS || message.member.roles.cache.some(role => role.id === '332343869163438080'))) {
+  if (!(message.member.permissions.serialize().KICK_MEMBERS || message.member.permissions.serialize().BAN_MEMBERS || message.member.roles.cache.some(role => role.id === '332343869163438080'))) {
```
Lets keep this premium for now, as I haven't built ability to set up investigations channel just yet, thus it's being done manually in DB for premium servers right now.

This functionality will be free, but I'll keep toxicity monitoring locked due to performance constraints and API limits.
```suggestion
  if (server.isPremium && !(message.member.permissions.serialize().KICK_MEMBERS || message.member.permissions.serialize().BAN_MEMBERS || message.member.roles.cache.some(role => role.id === '332343869163438080'))) {
```

lost-in-action replied:
Ah yes of course I didn't mean to push this haha

Jasius commented on src/utils.js:
```diff
@@ -51,6 +51,7 @@ const getDaysSince = (current, previous) => {
 }
 
 const collectCommandAnalytics = async (commandName, subCommandName) => {
+    return; // SHUT UP AND STOP BOTHERING ME WITH EVENTS.JS OH MY GOD JUST SHUT UP
```
Haha, you can just create an event.js file with an empty JSON array and it will work fine. Sorry for that, I should update it so that it creates file on its own, instead of crashing if there isn't one.
```suggestion
 
```

lost-in-action replied:
Oh no I didn't meant to commit this, I forgot to uncheck this lol, I did what you said before but sometimes the file kept getting cleared so I was getting mad haha

Jasius reviewed (commented):
Hey, I've been on kayaking trip so haven't had time to do a review up until now. Thank you for this thoughtful PR, I've ran through this code once and added some changes(mostly copywriting suggestions)

Also tried running this on a server without community rules set up and it didn't seem to fail gracefuly, just popped off some messaging and errors in the logs.
![image](https://user-images.githubusercontent.com/11031051/131254642-c82becd9-b722-4a89-b568-38fce043a9cc.png)

Going to do more testing now.

Jasius commented:
Things that I've noticed so far

- [ ] It errors out when trying to report on a server without community rules
- [ ] Thread doesn't get archived after the action as per user observation 
![image](https://user-images.githubusercontent.com/11031051/131255145-6a3a3a19-e78f-4c7f-bfd5-39976d1e64fb.png)

- [ ] It seems to create a new thread instead of re-opening the previous user thread.

Jasius commented:
> I didn't introduce the feature that hides and then reinstate the reported message, I'm not sure if that would be what same actions as flagging means.

I don't think that this is necessary, it's only necessary to delete the message in case moderators click the approval action.

lost-in-action commented:
> > I didn't introduce the feature that hides and then reinstate the reported message, I'm not sure if that would be what same actions as flagging means.
> 
> I don't think that this is necessary, it's only necessary to delete the message in case moderators click the approval action. It seems that message doesn't get deleted upon approval action, nor user gets muted.

Yes I think that's a bad idea to flag message once reported, it would be open to abuse, and reported users shouldn't be aware about it (some trolls would be like "who reported me?" and try to figure it out, would be a mess)

lost-in-action commented on src/db/index.js:
```diff
@@ -283,5 +285,40 @@ module.exports = {
+  saveBlockedReportUser: async function(userId, username) {
+    try {
+      const filter = { userId };
+      const result = await BlockedReport.findOneAndUpdate(filter, { $setOnInsert: {
+          userId,
+          username
```
> Hey, I've been on kayaking trip so haven't had time to do a review up until now. Thank you for this thoughtful PR, I've ran through this code once and added some changes(mostly copywriting suggestions)
> 
> Also tried running this on a server without community rules set up and it didn't seem to fail gracefuly, just popped off some messaging and errors in the logs.
> ![image](https://user-images.githubusercontent.com/11031051/131254642-c82becd9-b722-4a89-b568-38fce043a9cc.png)
> 
> Going to do more testing now.

I didn't implement that, currently it uses "fetchRules()" but we need something like an interface that fallback to custom rules if it's not a community server: like we need a new command to set up custom rules in the bot, or maybe make the bot use a #rules channel and parse rules from there, need to think more about it

Jasius replied:
> but we need something like an interface that fallback to custom rules if it's not a community server:

I think for MVP it's better to just leave servers without rules unsupported, as /rules command doesn't support non community servers either so that would also be consistent with it.

lost-in-action commented:
I fixed these issues:
1. Now errors if the server isn't community (check if the guild has a rule channel)
2. The thread now gets archived after review. The only problem would be if there are multiple reports pending, but that's a trivial issue since mods can just unarchive
3. Fixed (I hope) the issue where it would create duplicate threads

> It requires you to click rules confirmation button twice to make message update, even though report gets sent after one click.

I think I get what you mean
When the dropdown list goes away, it updates the interaction in accordance to what you have selected, components get disabled as soon as the interaction updates; so it really looks like you are clicking on the button, but it was disabled before, because the dropdown list updates first.

Rules update once the dropdown list goes away (you can click anywhere except on the dropdown itself to make it disappear), afterwards you can click one of the buttons
